sample-tracking-status-overview-app/src/main/groovy/life/qbic/portal/sampletracking/resource/ResourceService.groovy

```diff
@@ -0,0 +1,113 @@
+package life.qbic.portal.sampletracking.resource
+
+import life.qbic.portal.sampletracking.communication.Channel
+import life.qbic.portal.sampletracking.communication.Subscriber
+import life.qbic.portal.sampletracking.communication.Topic
+
+/**
+ * <b>A service that holds a resource of type T and communicates changes to its content</b>
+ *
+ * <p>This service acts as a channel where publishers can publish changes to data
+ * and subscribers are notified of changes. The data that is published is stored in the service.</p>
+ *
+ * @param <T>  the type of data to be communicated using this service
+ * @since 1.0.0
+ */
+abstract class ResourceService<T> {
+    protected final List<? extends T> content
+    protected final Map<Topic, Channel<T>> channels
+
+    /**
+     * Creates a resource service with an empty content.
+     */
+    ResourceService() {
+        this.content = new ArrayList<>() //generics inferred from field
+        channels = new HashMap<>()
+    }
+
+    /**
+     * Subscribes to service update events. Update events are emitted by the service when new
+     * resource items are added, removed or the resource has refreshed.
+     *
+     * @param subscriber The subscriber to register for update events
+     * @param topic the topic to subscribe to
+     */
+    void subscribe(Subscriber<T> subscriber, Topic topic) {
+        if( channels.containsKey(topic)) {
+            channels.get(topic).subscribe(subscriber)
+        } else {
+            Channel<T> topicChannel = new Channel<T>()
+            channels.put(topic, topicChannel)
+            topicChannel.subscribe(subscriber)
+        }
```

Comment:
I would assume that the subscriber cannot define which topics are implemented.
Therefore I would expect there to be a method to add new topics

Reply:
Yes you are right. We need to decide on what to do then. We could throw an IllegalArgument exception.

Reply:
57d525f